File: apps/proxy/internal/proxy/attempts.go
package proxy

import (
	"context"
	"net/http"
	"strings"
	"time"

	"github.com/opendum/opendum/apps/proxy/internal/auth"
	appdb "github.com/opendum/opendum/apps/proxy/internal/db"
	"github.com/opendum/opendum/apps/proxy/internal/providers"
)

type accountRotationRunner interface {
	getNextAvailableAccount(context.Context, string, string, *string, []string, auth.AccountAccess) (*appdb.ProviderAccount, bool, error)
	getNextSharedAccount(context.Context, string, string, *string, []string) (*appdb.ProviderAccount, bool, error)
	reserveRoamingPoint(context.Context, string) (*pointReservation, bool, error)
	refundRoamingPoint(context.Context, *pointReservation)
	bumpAccountRequestCount(context.Context, string, time.Time)
	makeProviderRequest(context.Context, appdb.ProviderAccount, map[string]any, bool) (*http.Response, error)
	markAccountFailed(context.Context, string, string, int, string) time.Time
	markAccountUsageLimited(context.Context, string, string, time.Time, time.Time)
	logUsage(context.Context, usageParams)
	isVisionModel(string) bool
	isToolCallModel(string) bool
	canAccountUseModel(appdb.ProviderAccount, string) bool
}

type delayedAccountFailure struct {
	account    appdb.ProviderAccount
	statusCode int
	message    string
}

type accountAttempt struct {
	account *appdb.ProviderAccount
	roaming *pointReservation
}

func (s *Service) executeWithAccountRotation(ctx context.Context, r *http.Request, cfg endpointAdapter, parsed parsedEndpointRequest, authResult auth.Result, validation auth.ModelValidationResult, forced *appdb.ProviderAccount, startMS int64) (*appdb.ProviderAccount, *http.Response, int64, int64, []accountRotationFailure, *pointReservation, *routeError) {
	return executeAccountRotation(s, ctx, r, cfg, parsed, authResult, validation, forced, startMS)
}

func executeAccountRotation(runner accountRotationRunner, ctx context.Context, r *http.Request, cfg endpointAdapter, parsed parsedEndpointRequest, authResult auth.Result, validation auth.ModelValidationResult, forced *appdb.ProviderAccount, startMS int64) (*appdb.ProviderAccount, *http.Response, int64, int64, []accountRotationFailure, *pointReservation, *routeError) {
	tried := []string{}
	sharedTried := []string{}
	useShared := false
	recoverableFailures := []accountRotationFailure{}
	var lastFailure *routeError
	var delayedFinalFailure *delayedAccountFailure
	accountConfigured := false

	for {
		attempt := accountAttempt{account: forced}
		if attempt.account == nil {
			selected, configured, err := nextAttemptAccount(runner, ctx, authResult, validation, tried, sharedTried, useShared)
			if err != nil {
				return nil, nil, 0, 0, recoverableFailures, nil, &routeError{Status: http.StatusInternalServerError, Message: "Internal server error", Type: "api_error"}
			}
			accountConfigured = accountConfigured || configured
			attempt.account = selected
		}

		if attempt.account == nil && !useShared && forced == nil && authResult.RoamingEnabled {
			useShared = true
			continue
		}

		if attempt.account == nil {
			if lastFailure != nil && delayedFinalFailure != nil {
				runner.markAccountFailed(ctx, delayedFinalFailure.account.ID, validation.Model, delayedFinalFailure.statusCode, delayedFinalFailure.message)
			}
			if len(tried)+len(sharedTried) == 0 {
				if accountConfigured {
					return nil, nil, 0, 0, recoverableFailures, nil, &routeError{Status: http.StatusServiceUnavailable, Message: "This model is temporarily unavailable. Please try again later.", Type: "api_error"}
				}
				return nil, nil, 0, 0, recoverableFailures, nil, &routeError{Status: cfg.NoAccountsStatusCode, Message: "No active accounts available for this model. Please add an account in the dashboard.", Type: "configuration_error"}
			}
			if lastFailure != nil {
				return nil, nil, 0, 0, recoverableFailures, nil, lastFailure
			}
			return nil, nil, 0, 0, recoverableFailures, nil, &routeError{Status: http.StatusServiceUnavailable, Message: "No available accounts for this request.", Type: "api_error"}
		}
		if useShared && forced == nil {
			points, allowed, err := runner.reserveRoamingPoint(ctx, authResult.UserID)
			if err != nil {
				return nil, nil, 0, 0, recoverableFailures, nil, &routeError{Status: http.StatusInternalServerError, Message: "Internal server error", Type: "api_error"}
			}
			if !allowed {
				code := "insufficient_points"
				return nil, nil, 0, 0, recoverableFailures, nil, &routeError{Status: http.StatusPaymentRequired, Message: "Insufficient points. Please add more points to continue.", Type: "insufficient_quota", Code: &code}
			}
			attempt.roaming = points
		}

		if useShared {
			sharedTried = append(sharedTried, attempt.account.ID)
		} else {
			tried = append(tried, attempt.account.ID)
		}
		if forced != nil {
			go runner.bumpAccountRequestCount(context.Background(), attempt.account.ID, time.Now())
		}
		payload := cfg.Build(parsed, validation.Model, parsed.Stream, sessionID(r))
		if !runner.isVisionModel(validation.Model) {
			stripImageContent(payload)
		}
		if !runner.isToolCallModel(validation.Model) {
			stripToolCallParameters(payload)
		}
		requestStart := time.Now().UnixMilli()
		upstreamFirstResponseMS := int64(0)
		attemptCtx := providers.WithUpstreamResponseStartRecorder(ctx, func(at time.Time) {
			if upstreamFirstResponseMS == 0 {
				upstreamFirstResponseMS = at.UnixMilli()
			}
		})
		resp, err := runner.makeProviderRequest(attemptCtx, *attempt.account, payload, parsed.Stream)
		if upstreamFirstResponseMS == 0 && resp != nil {
			upstreamFirstResponseMS = time.Now().UnixMilli()
		}
		if err != nil {
			delayedFinalFailure = nil
			status := http.StatusInternalServerError
			message := err.Error()
			detailed := buildAccountErrorMessage(message, accountErrorContext{Model: validation.Model, Provider: attempt.account.Provider, Endpoint: endpointPath(cfg.Endpoint), Messages: parsed.MessagesForError, Parameters: parsed.ParamsForError})
			failedAt := runner.markAccountFailed(ctx, attempt.account.ID, validation.Model, status, detailed)
			runner.logUsage(ctx, usageParams{UserID: authResult.UserID, ProviderAccountID: attempt.account.ID, ProxyAPIKeyID: authResult.APIKeyID, Model: validation.Model, StatusCode: status, DurationMS: int(time.Now().UnixMilli() - startMS), Provider: attempt.account.Provider})
			lastFailure = &routeError{Status: status, Message: message, Type: "api_error"}
			if attempt.roaming != nil {
				runner.refundRoamingPoint(context.Background(), attempt.roaming)
			}
			if forced == nil {
				recoverableFailures = append(recoverableFailures, accountRotationFailure{AccountID: attempt.account.ID, FailedAt: failedAt})
				continue
			}
			return nil, nil, 0, 0, recoverableFailures, nil, lastFailure
		}

		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
			bodyText := readBodyLimit(resp.Body, 1<<20)
			_ = resp.Body.Close()
			var failedAt time.Time
			delayedFinalFailure = nil
			if resp.StatusCode != http.StatusRequestTimeout {
				detailed := buildAccountErrorMessage(bodyText, accountErrorContext{Model: validation.Model, Provider: attempt.account.Provider, Endpoint: endpointPath(cfg.Endpoint), Messages: parsed.MessagesForError, Parameters: parsed.ParamsForError})
				if forced == nil && isAntigravityResourceExhausted(attempt.account.Provider, resp.StatusCode, bodyText) {
					delayedFinalFailure = &delayedAccountFailure{account: *attempt.account, statusCode: resp.StatusCode, message: detailed}
				} else {
					failedAt = runner.markAccountFailed(ctx, attempt.account.ID, validation.Model, resp.StatusCode, detailed)
					if disabledUntil, ok := codexUsageLimitDisabledUntil(attempt.account.Provider, resp.StatusCode, bodyText, failedAt); ok {
						runner.markAccountUsageLimited(ctx, attempt.account.ID, validation.Model, disabledUntil, failedAt)
					}
				}
			}
			runner.logUsage(ctx, usageParams{UserID: authResult.UserID, ProviderAccountID: attempt.account.ID, ProxyAPIKeyID: authResult.APIKeyID, Model: validation.Model, StatusCode: resp.StatusCode, DurationMS: int(time.Now().UnixMilli() - startMS), Provider: attempt.account.Provider})
			message, typ := sanitizedProxyError(resp.StatusCode, bodyText)
			lastFailure = &routeError{Status: resp.StatusCode, Message: message, Type: typ}
			if attempt.roaming != nil {
				runner.refundRoamingPoint(context.Background(), attempt.roaming)
			}
			if shouldRotate(resp.StatusCode) && forced == nil {
				if !failedAt.IsZero() {
					recoverableFailures = append(recoverableFailures, accountRotationFailure{AccountID: attempt.account.ID, FailedAt: failedAt})
				}
				continue
			}
			return nil, nil, 0, 0, recoverableFailures, nil, lastFailure
		}

		return attempt.account, resp, requestStart, upstreamFirstResponseMS, recoverableFailures, attempt.roaming, nil
	}
}

func nextAttemptAccount(runner accountRotationRunner, ctx context.Context, authResult auth.Result, validation auth.ModelValidationResult, tried, sharedTried []string, useShared bool) (*appdb.ProviderAccount, bool, error) {
	if useShared {
		return runner.getNextSharedAccount(ctx, authResult.UserID, validation.Model, validation.Provider, sharedTried)
	}
	return runner.getNextAvailableAccount(ctx, authResult.UserID, validation.Model, validation.Provider, tried, auth.AccountAccess{Mode: authResult.AccountAccessMode, Accounts: authResult.AccountAccessList})
}

func (s *Service) isVisionModel(model string) bool {
	if s.registry == nil {
		return true
	}
	return s.registry.IsVisionModel(model)
}

func (s *Service) isToolCallModel(model string) bool {
	if s.registry == nil {
		return true
	}
	return s.registry.IsToolCallModel(model)
}

func (s *Service) canAccountUseModel(account appdb.ProviderAccount, model string) bool {
	if s.registry == nil {
		return true
	}
	rule, ok := s.registry.ProviderAccessRule(model, account.Provider)
	if !ok || !accountAccessRuleRestrictsTier(rule.MinTier, rule.AllowedTiers) {
		return true
	}
	return accountTierSatisfiesRule(quotaFallbackTier(account), rule.MinTier, rule.AllowedTiers)
}

func accountTierSatisfiesRule(accountTier, minTier string, allowedTiers []string) bool {
	normalizedAccountTier := normalizeAccountTierAlias(accountTier)
	if len(allowedTiers) > 0 {
		for _, tier := range allowedTiers {
			if normalizeAccountTierAlias(tier) == normalizedAccountTier {
				return true
			}
		}
		return false
	}

	required := strings.ToLower(strings.TrimSpace(minTier))
	if required == "" || required == "free" {
		return true
	}
	return normalizedAccountTier == normalizeAccountTierAlias(required)
}

func accountAccessRuleRestrictsTier(minTier string, allowedTiers []string) bool {
	if len(allowedTiers) > 0 {
		return true
	}
	required := normalizeAccountTierAlias(minTier)
	return required != "" && required != "free"
}

func normalizeAccountTierAlias(tier string) string {
	normalized := strings.ToLower(strings.ReplaceAll(strings.TrimSpace(tier), "_", "-"))
	if normalized == "pro-plus" || normalized == "proplus" {
		return "pro+"
	}
	if normalized == "free-tier" || normalized == "free-limited-copilot" {
		return "free"
	}
	if normalized == "education" || normalized == "educational" || normalized == "edu" || normalized == "free-educational-quota" {
		return "student"
	}
	return normalized
}

func sessionID(r *http.Request) string {
	if value := r.Header.Get("session_id"); value != "" {
		return value
	}
	return r.Header.Get("x-session-id")
}

File: apps/proxy/src/core/attempts.ts
import type { AuthResult, ModelValidationResult } from "../auth/types.js";
import type { ProviderAccount } from "../db/index.js";
import type { UpstreamResponse } from "../providers/types.js";
import { buildAccountErrorMessage, codexUsageLimitDisabledUntil, endpointPath, isAntigravityResourceExhausted, prefixWithProvider, readBodyLimit, sanitizedProxyError, shouldRotate } from "./errors.js";
import { stripImageContent, stripToolCallParameters } from "./content.js";
import type { AccountRotationFailure, EndpointAdapter, ParsedEndpointRequest, RouteError } from "./types.js";
import type { PointReservation } from "./points.js";

export interface RotationRunner {
  getNextAvailableAccount(userID: string, model: string, provider: string | null, exclude: string[], excludeProviders: string[], accountAccess: { mode: string; accounts: string[] }, sessionID: string): Promise<[ProviderAccount | null, boolean, Error | null]>;
  getNextSharedAccount(userID: string, model: string, provider: string | null, exclude: string[], excludeProviders: string[]): Promise<[ProviderAccount | null, boolean, Error | null]>;
  reserveRoamingPoint(userID: string): Promise<[PointReservation | null, boolean, Error | null]>;
  refundRoamingPoint(reservation: PointReservation | null): Promise<void>;
  bumpAccountRequestCount(accountID: string, usedAt: Date): Promise<void>;
  makeProviderRequest(account: ProviderAccount, payload: Record<string, unknown>, stream: boolean, ctx: Record<string, unknown>): Promise<UpstreamResponse>;
  markAccountFailed(accountID: string, model: string, status: number, message: string): Promise<Date>;
  markAccountUsageLimited(accountID: string, model: string, disabledUntil: Date, failedAt: Date): Promise<void>;
  logUsage(params: { userId: string; providerAccountId: string; proxyApiKeyId: string; model: string; statusCode: number; durationMS: number; provider: string }): Promise<void>;
  isVisionModel(model: string): boolean;
  isToolCallModel(model: string): boolean;
}

export interface RotationResult {
  account: ProviderAccount | null;
  response: UpstreamResponse | null;
  requestStartMS: number;
  upstreamFirstResponseMS: number;
  rotationFailures: AccountRotationFailure[];
  roaming: PointReservation | null;
  error: RouteError | null;
}

export async function executeAccountRotation(
  runner: RotationRunner,
  cfg: EndpointAdapter,
  parsed: ParsedEndpointRequest,
  authResult: AuthResult,
  validation: ModelValidationResult,
  forced: ProviderAccount | null,
  startMS: number,
  sessionID: string,
): Promise<RotationResult> {
  const tried: string[] = [];
  const sharedTried: string[] = [];
  const excludedProviders: string[] = [];
  let useShared = false;
  const recoverableFailures: AccountRotationFailure[] = [];
  let lastFailure: RouteError | null = null;
  let delayedFinalFailure: { account: ProviderAccount; statusCode: number; message: string } | null = null;
  let accountConfigured = false;

  for (;;) {
    let attempt: { account: ProviderAccount | null; roaming: PointReservation | null } = { account: forced, roaming: null };
    if (attempt.account === null) {
      let selected: ProviderAccount | null = null;
      let configured = false;
      let err: Error | null = null;
      if (useShared) {
        [selected, configured, err] = await runner.getNextSharedAccount(authResult.userId, validation.model, validation.provider, sharedTried, excludedProviders);
      } else {
        [selected, configured, err] = await runner.getNextAvailableAccount(authResult.userId, validation.model, validation.provider, tried, excludedProviders, { mode: authResult.accountAccessMode, accounts: authResult.accountAccessList }, sessionID);
      }
      if (err) {
        return { account: null, response: null, requestStartMS: 0, upstreamFirstResponseMS: 0, rotationFailures: recoverableFailures, roaming: null, error: { status: 500, message: "Internal server error", type: "api_error", param: null, code: null, retryAfter: null, retryAfterMS: null, accountID: "" } };
      }
      accountConfigured = accountConfigured || configured;
      attempt.account = selected;
    }

    if (attempt.account === null && !useShared && forced === null && authResult.roamingEnabled) {
      useShared = true;
      continue;
    }

    if (attempt.account === null) {
      if (lastFailure && delayedFinalFailure) {
        await runner.markAccountFailed(delayedFinalFailure.account.id, validation.model, delayedFinalFailure.statusCode, delayedFinalFailure.message);
      }
      if (tried.length + sharedTried.length === 0) {
        if (accountConfigured) {
          return { account: null, response: null, requestStartMS: 0, upstreamFirstResponseMS: 0, rotationFailures: recoverableFailures, roaming: null, error: { status: 503, message: "This model is temporarily unavailable. Please try again later.", type: "api_error", param: null, code: null, retryAfter: null, retryAfterMS: null, accountID: "" } };
        }
        return { account: null, response: null, requestStartMS: 0, upstreamFirstResponseMS: 0, rotationFailures: recoverableFailures, roaming: null, error: { status: cfg.noAccountsStatusCode, message: "No active accounts available for this model. Please add an account in the dashboard.", type: "configuration_error", param: null, code: null, retryAfter: null, retryAfterMS: null, accountID: "" } };
      }
      if (lastFailure) {
        return { account: null, response: null, requestStartMS: 0, upstreamFirstResponseMS: 0, rotationFailures: recoverableFailures, roaming: null, error: lastFailure };
      }
      return { account: null, response: null, requestStartMS: 0, upstreamFirstResponseMS: 0, rotationFailures: recoverableFailures, roaming: null, error: { status: 503, message: "No available accounts for this request.", type: "api_error", param: null, code: null, retryAfter: null, retryAfterMS: null, accountID: "" } };
    }
    if (useShared && forced === null) {
      const [points, allowed, err] = await runner.reserveRoamingPoint(authResult.userId);
      if (err) {
        return { account: null, response: null, requestStartMS: 0, upstreamFirstResponseMS: 0, rotationFailures: recoverableFailures, roaming: null, error: { status: 500, message: "Internal server error", type: "api_error", param: null, code: null, retryAfter: null, retryAfterMS: null, accountID: "" } };
      }
      if (!allowed) {
        return { account: null, response: null, requestStartMS: 0, upstreamFirstResponseMS: 0, rotationFailures: recoverableFailures, roaming: null, error: { status: 402, message: "Insufficient points. Please add more points to continue.", type: "insufficient_quota", param: null, code: "insufficient_points", retryAfter: null, retryAfterMS: null, accountID: "" } };
      }
      attempt.roaming = points;
    }

    if (useShared) {
      sharedTried.push(attempt.account.id);
    } else {
      tried.push(attempt.account.id);
    }
    if (forced !== null) {
      void runner.bumpAccountRequestCount(attempt.account.id, new Date());
    }
    const payload = cfg.build(parsed, validation.model, parsed.stream, sessionID);
    if (!runner.isVisionModel(validation.model)) {
      stripImageContent(payload);
    }
    if (!runner.isToolCallModel(validation.model)) {
      stripToolCallParameters(payload);
    }
    const requestStart = Date.now();
    let upstreamFirstResponseMS = 0;
    const attemptCtx: Record<string, unknown> = {
      recordResponseStart: (at: Date) => {
        if (upstreamFirstResponseMS === 0) upstreamFirstResponseMS = at.getTime();
      },
    };
    let resp: UpstreamResponse;
    try {
      resp = await runner.makeProviderRequest(attempt.account, payload, parsed.stream, attemptCtx);
    } catch (error) {
      const err = error as Error;
      delayedFinalFailure = null;
      const status = 500;
      const message = err.message;
      const detailed = buildAccountErrorMessage(message, { model: validation.model, provider: attempt.account.provider, endpoint: endpointPath(cfg.endpoint), messages: parsed.messagesForError, parameters: parsed.paramsForError });
      const failedAt = await runner.markAccountFailed(attempt.account.id, validation.model, status, detailed);
      await runner.logUsage({ userId: authResult.userId, providerAccountId: attempt.account.id, proxyApiKeyId: authResult.apiKeyId, model: validation.model, statusCode: status, durationMS: Date.now() - startMS, provider: attempt.account.provider });
      lastFailure = { status, message: prefixWithProvider(attempt.account.provider, message), type: "api_error", param: null, code: null, retryAfter: null, retryAfterMS: null, accountID: attempt.account.id };
      if (attempt.roaming) {
        await runner.refundRoamingPoint(attempt.roaming);
      }
      if (forced === null) {
        recoverableFailures.push({ accountId: attempt.account.id, failedAt });
        continue;
      }
      return { account: null, response: null, requestStartMS: 0, upstreamFirstResponseMS: 0, rotationFailures: recoverableFailures, roaming: null, error: lastFailure };
    }
    if (upstreamFirstResponseMS === 0 && resp) {
      upstreamFirstResponseMS = Date.now();
    }

    if (resp.status < 200 || resp.status >= 300) {
      const bodyText = await readBodyLimit(resp.body, 1 << 20);
      let failedAt: Date | null = null;
      delayedFinalFailure = null;
      const badRequestFromProvider = resp.status === 400;
      const fallbackAcrossProviders = badRequestFromProvider && forced === null && validation.provider === null;
      if (resp.status !== 408 && !badRequestFromProvider) {
        const detailed = buildAccountErrorMessage(bodyText, { model: validation.model, provider: attempt.account.provider, endpoint: endpointPath(cfg.endpoint), messages: parsed.messagesForError, parameters: parsed.paramsForError });
        if (forced === null && isAntigravityResourceExhausted(attempt.account.provider, resp.status, bodyText)) {
          delayedFinalFailure = { account: attempt.account, statusCode: resp.status, message: detailed };
        } else {
          failedAt = await runner.markAccountFailed(attempt.account.id, validation.model, resp.status, detailed);
          const [disabledUntil, ok] = codexUsageLimitDisabledUntil(attempt.account.provider, resp.status, bodyText, failedAt);
          if (ok) {
            await runner.markAccountUsageLimited(attempt.account.id, validation.model, disabledUntil, failedAt);
          }
        }
      }
      await runner.logUsage({ userId: authResult.userId, providerAccountId: attempt.account.id, proxyApiKeyId: authResult.apiKeyId, model: validation.model, statusCode: resp.status, durationMS: Date.now() - startMS, provider: attempt.account.provider });
      const [message, typ] = sanitizedProxyError(resp.status, bodyText);
      lastFailure = { status: resp.status, message: prefixWithProvider(attempt.account.provider, message), type: typ, param: null, code: null, retryAfter: null, retryAfterMS: null, accountID: attempt.account.id };
      if (attempt.roaming) {
        await runner.refundRoamingPoint(attempt.roaming);
      }
      if (fallbackAcrossProviders) {
        excludedProviders.push(attempt.account.provider);
        continue;
      }
      if (shouldRotate(resp.status) && forced === null) {
        if (failedAt) {
          recoverableFailures.push({ accountId: attempt.account.id, failedAt });
        }
        continue;
      }
      return { account: null, response: null, requestStartMS: 0, upstreamFirstResponseMS: 0, rotationFailures: recoverableFailures, roaming: null, error: lastFailure };
    }

    return { account: attempt.account, response: resp, requestStartMS: requestStart, upstreamFirstResponseMS, rotationFailures: recoverableFailures, roaming: attempt.roaming, error: null };
  }
}

export function sessionID(request: Request): string {
  const value = request.headers.get("session_id");
  if (value) return value;
  return request.headers.get("x-session-id") ?? "";
}
